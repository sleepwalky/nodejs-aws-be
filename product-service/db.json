[
  {
    "count": 4,
    "description": "Дополнение",
    "id": "7567ec4b-b10c-48c5-9345-fc73c48a80aa",
    "price": 87,
    "img": "https://igraj.by/upload/iblock/f5a/f5a48e0762001ad9be2371423df702ab.jpg",
    "title": "Крылья: Птицы Европы"
  },
  {
    "count": 6,
    "description": "Short Product Description3",
    "id": "7567ec4b-b10c-48c5-9345-fc73c48a80a0",
    "price": 175,
    "img": "https://igraj.by/upload/Sh/imageCache/6ac/511/2f4c76754a96582d6db19f00e72a312e.jpg",
    "title": "Маракайбо"
  },
  {
    "count": 7,
    "description": "Short Product Description2",
    "id": "7567ec4b-b10c-48c5-9345-fc73c48a80a2",
    "price": 17,
    "img": "https://igraj.by/upload/Sh/imageCache/86d/efa/f6f4124a393a4c6c0a641c0625ae4368.jpg",
    "title": "Соображарий Супер"
  },
  {
    "count": 12,
    "description": "Short Product Description7",
    "id": "7567ec4b-b10c-48c5-9345-fc73c48a80a1",
    "price": 116,
    "img": "https://igraj.by/upload/Sh/imageCache/259/95a/420580760fcf1a89aac7c1929b2860f6.jpg",
    "title": "Цивилизация Сида Мейера: Новый рассвет. Терра инкогнита"
  },
  {
    "count": 7,
    "description": "Short Product Description2",
    "id": "7567ec4b-b10c-48c5-9345-fc73c48a80a3",
    "price": 31,
    "img": "https://igraj.by/upload/Sh/imageCache/47c/d51/cae2ade43f0046c6da31e6dac6c8840e.jpg",
    "title": "Бэнг! Звёздные Войны"
  },
  {
    "count": 8,
    "description": "Short Product Description4",
    "id": "7567ec4b-b10c-48c5-9345-fc73348a80a6",
    "price": 420,
    "img": "https://igraj.by/upload/Sh/imageCache/5cb/8df/3bb1ac23ac7ffa8255a6bff45d3fb400.jpg",
    "title": "Немезида / Nemesis"
  },
  {
    "count": 2,
    "description": "Short Product Descriptio1",
    "id": "7567ec4b-b10c-48c5-9445-fc73c48a80a5",
    "price": 175,
    "img": "https://igraj.by/upload/Sh/imageCache/633/662/aeb2601974000ce9a38f8446f233e766.jpg",
    "title": "Остров Духов"
  },
  {
    "count": 3,
    "description": "Short Product Description7",
    "id": "7567ec4b-b10c-45c5-9345-fc73c48a80a4",
    "price": 105,
    "img": "https://igraj.by/upload/Sh/imageCache/b1e/936/7108af77ff6beaf9c3042c8cf0a280e1.png",
    "title": "Ярость Дракулы"
  }
]
