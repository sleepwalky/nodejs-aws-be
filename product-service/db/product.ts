import { Client, ClientConfig } from 'pg';
import { Product } from '../models/Product';

const { PG_HOST, PG_PORT, PG_DATABASE, PG_USERNAME, PG_PASSWORD } = process.env;
const dbOptions: ClientConfig = {
  host: PG_HOST,
  port: parseInt(PG_PORT, 10),
  database: PG_DATABASE,
  user: PG_USERNAME,
  password: PG_PASSWORD,
  ssl: {
    rejectUnauthorized: false, // to avoid warring in this example
  },
  connectionTimeoutMillis: 5000, // time in millisecond for termination of the database query
};

export const getProductsFromDb = async () => {
  const client = new Client(dbOptions);
  await client.connect();

  try {
    const { rows: products } = await client.query(
      `select * from products p inner join stocks s on p.id = s.product_id `
    );
    console.log(products);
    return products;
  } catch (err) {
    console.error('Error during database request executing:', err);
    throw err;
  } finally {
    client.end();
  }
};

export const getProductByIdFromDb = async (id: string) => {
  const client = new Client(dbOptions);
  await client.connect();

  try {
    const { rows: products } = await client.query(
      `select * from products p inner join stocks s on p.id = s.product_id where p.id = '${id}'`
    );
    console.log(products);
    return products[0];
  } catch (err) {
    console.error('Error during database request executing:', err);
    throw err;
  } finally {
    client.end();
  }
};

export const createProductDb = async ({
  description,
  title,
  img,
  count,
  price,
}: Product) => {
  const client = new Client(dbOptions);
  await client.connect();

  try {
    await client.query('BEGIN');
    const queryText =
      'INSERT INTO products(title, description, price, img) VALUES($1, $2, $3, $4) RETURNING id';
    const insertProductValues = [title, description, price, img];
    console.log(insertProductValues);
    const res = await client.query(queryText, insertProductValues);
    console.log('succesfull insert into products');
    const insertStocksText =
      'INSERT INTO stocks(product_id, count) VALUES ($1, $2)';
    const insertStocksValues = [res.rows[0].id, count];
    await client.query(insertStocksText, insertStocksValues);
    await client.query('COMMIT');
    return await getProductByIdFromDb(res.rows[0].id);
  } catch (e) {
    await client.query('ROLLBACK');
    console.log(e);
    throw e;
  } finally {
    client.end();
  }
};
