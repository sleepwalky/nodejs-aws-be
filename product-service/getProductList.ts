import { APIGatewayProxyHandler } from 'aws-lambda';
import 'source-map-support/register';
import { commonHeaders } from './commonHeaders';
import { getProductsFromDb } from './services';

export const getProductList: APIGatewayProxyHandler = async (
  event,
  _context
) => {
  try {
    const products = await getProductsFromDb();
    return {
      statusCode: 200,
      headers: commonHeaders,
      body: JSON.stringify(
        {
          items: products,
        },
        null,
        2
      ),
    };
  } catch (error) {
    return {
      statusCode: 500,
      headers: commonHeaders,
      body: JSON.stringify({
        error: {
          message: "Couldn't load products from DB",
        },
      }),
    };
  }
};
