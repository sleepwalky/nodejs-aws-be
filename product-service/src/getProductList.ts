import { APIGatewayProxyHandler } from 'aws-lambda';
import 'source-map-support/register';
import { commonHeaders } from '../../commonHeaders';
import { getProductsFromDb } from '../db/product';

export const getProductList: APIGatewayProxyHandler = async (
  event,
  _context
) => {
  try {
    console.log(event);
    const products = await getProductsFromDb();
    return {
      statusCode: 200,
      headers: commonHeaders,
      body: JSON.stringify(
        {
          items: products,
        },
        null,
        2
      ),
    };
  } catch (error) {
    return {
      statusCode: 500,
      headers: commonHeaders,
      body: JSON.stringify({
        error: {
          message: "Couldn't load products from DB",
        },
      }),
    };
  }
};
