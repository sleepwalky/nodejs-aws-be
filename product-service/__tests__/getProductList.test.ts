import { getProductList } from '../getProductList';
import { getProductsFromDb } from '../db/product';

jest.mock('../db/product', () => ({
  __esModule: true,
  getProductsFromDb: jest.fn(),
}));

describe('getProductList', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should return product list', async () => {
    (getProductsFromDb as jest.Mock).mockResolvedValue([
      { id: '1' },
      { id: '2' },
    ]);
    const result = await getProductList({} as any, {} as any, () => {});

    expect(result).toMatchSnapshot();
  });

  it('should return 500 if db throws', async () => {
    (getProductsFromDb as jest.Mock).mockRejectedValue(null);
    const result = await getProductList(
      {
        pathParameters: {
          productId: '123',
        },
      } as any,
      {} as any,
      () => {}
    );

    expect(result).toMatchSnapshot();
  });
});
