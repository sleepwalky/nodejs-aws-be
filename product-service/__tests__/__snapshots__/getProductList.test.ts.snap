// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProductList should return 500 if db throws 1`] = `
Object {
  "body": "{\\"error\\":{\\"message\\":\\"Couldn't load products from DB\\"}}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 500,
}
`;

exports[`getProductList should return product list 1`] = `
Object {
  "body": "{
  \\"items\\": [
    {
      \\"id\\": \\"1\\"
    },
    {
      \\"id\\": \\"2\\"
    }
  ]
}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;
